### ISSUE TITLE
[Documentation suggestion] Provide executable examples and link to them

### ISSUE THREAD

**Issue by martinmoene** [philsquared/Catch#1037]
Readers of Catch' documentation may benefit from readily available, complete and executable examples that show a limited and coherent set of features and steps needed for compilation. The tutorial and reference documentation can then link to relevant examples that the reader can [try, perhaps even online](https://wandbox.org/permlink/QzBURjkOxcFH99BF). 

Compilation of these examples must be integrated with the continuous integration process. As such the examples become [executable documentation](https://accu.org/index.php/journals/1871) geared towards the user.

For example:
```
// Assertions

// Compile:
// - g++ -I$(CATCH_SINGLE_INCLUDE) -o assertions assertions.cpp && assertions --success
// - cl -EHsc -I%CATCH_SINGLE_INCLUDE% assertions.cpp && assertions --success

// Let Catch provide a main():
#define CATCH_CONFIG_MAIN

#include "catch.hpp"

int one() {
    return 1;
}

TEST_CASE( "Assert that something is true (pass)", "[require]" ) {
    REQUIRE( one() == 1 );
}

TEST_CASE( "Assert that something is true (fail)", "[require]" ) {
    REQUIRE( one() == 2 );
}

TEST_CASE( "Assert that something is true (stop at first failure)", "[require]" ) {
    REQUIRE( one() == 2 );
    REQUIRE( one() == 1 );
}

TEST_CASE( "Assert that something is true (continue after failure)", "[check]" ) {
    CHECK(   one() == 2 );
    REQUIRE( one() == 1 );
}
```

Subjects that come to mind:
- test case (test_case)
- assertions (require, check)
- exceptions (require_nothrow, require_throws, require_throws_as, require_throws_with, require_throws_matches)
- matchers (require_that)
- floating point (approx)
- logging (capture, info, warn, fail, fail_check)
- fixtures (test_case_method)
- sections
- bdd (scenario, given, when, then, and_when, and_then)
- string conversions
- reporters
- event listeners
- supplying your own main()
- compile-time configuration

**Comment by horenmar** [comment]
Basically, you want parts of SelfTest to become self-contained and hosted on Wandbox or similar, correct?

**Comment by martinmoene** [comment]
Indeed, with the difference that SelfTest may be geared (too much) towards implementers, complete and overwhelming where mere users may need another approach to support their understanding.

Key:
- relatively simple
- complete program
- that compiles and executes and shows what's necessary for that
- and is at the tip if reader's fingers from the documentation

See for example the examples linked from [section Usage](https://github.com/martinmoene/lest#usage) of lest.

**Comment by horenmar** [comment]
Yeah, SelfTest's primary aim is to (mostly) guarantee that things didn't get broken by a change, secondary aim is a reference for someone who already knows Catch well and documentation for newbies is a very distant third.

Anyway, I can see some value in adding these, but honestly don't have much time (and what little I do, I'd like to use for different things), so this will likely either get put on a backburner, or someone else needs to put in a PR (hint, hint).

**Comment by martinmoene** [comment]
> I can see some value in adding these

is what I liked to know ;)

**Comment by martinmoene** [comment]
Plan (in progress):

Overall:
- [x] Create branch [`examples`](https://github.com/johannesmoene/Catch/tree/examples/)
- [x] Create folder `examples`
- [ ] Create several examples to get the gist, see below
- [ ] Link to examples from Tutorial and from Reference
- [ ] Publish branch for review feedback
- [ ] ...

Examples:
- [ ] 01-TestCase, single file
- [ ] 02-TestCase, multiple files
- [ ] 03-Assertions
- [ ] 04-...

**Comment by ahadji05** [comment]
I installed the Catch2 framework using the CMake installation and when I try to compile and run any example I get this error: `CATCH_SINGLE_INCLUDE: command not found`

this is how I compile an example:
`s$ g++ -std=c++11 -Wall -I$(CATCH_SINGLE_INCLUDE) -c 000-CatchMain.cpp`

what is the purpose of the macro `CATCH_SINGLE_INCLUDE` ?

**Comment by horenmar** [comment]
@ahadji05 It should contain path to the single-include version of `catch.hpp`.
